tabrepo/repository/evaluation_repository_collection.py

```diff
@@ -0,0 +1,175 @@
+from __future__ import annotations
+
+import copy
+from collections import defaultdict
+from typing import List
+
+import numpy as np
+import pandas as pd
+
+from .abstract_repository import AbstractRepository
+from .ground_truth_mixin import GroundTruthMixin
+from .evaluation_repository import EvaluationRepository
+from ..simulation.ensemble_selection_config_scorer import EnsembleSelectionConfigScorer
+from ..simulation.ground_truth import GroundTruth
+from ..simulation.simulation_context import ZeroshotSimulatorContext
+
+
+# TODO: WIP. This is not a fully functional class yet.
+class EvaluationRepositoryCollection(AbstractRepository, GroundTruthMixin):
+    """
+    Simple Repository class that implements core functionality related to
+    fetching model predictions, available datasets, folds, etc.
+    """
+    def __init__(
+            self,
+            repos: list[EvaluationRepository],
+            config_fallback: str = None,
+    ):
+        self.repos: list[EvaluationRepository] = repos
+        zeroshot_context = merge_zeroshot([repo._zeroshot_context for repo in self.repos])
+        self._ground_truth: GroundTruth = merge_ground_truth([repo._ground_truth for repo in self.repos])
+        super().__init__(zeroshot_context=zeroshot_context, config_fallback=config_fallback)
+
+        # DONE: Create AbstractRepository to avoid code-dupe
+        # TODO: raise exception if overlap in (dataset, fold, config)
```

Comment:
This code should work for generating the mapping and checking the intersection:

```python
import itertools
from typing import List, Tuple

from tabrepo import EvaluationRepository, load_repository
from tabrepo.contexts.context_artificial import load_repo_artificial


def _generate_dataset_fold_config_combinations(repos: List[EvaluationRepository]) -> List[List[Tuple[str, int, str]]]:
    # returns the combinations (dataset, fold, config) for each repository
    return [repo.metrics().index.tolist() for i, repo in enumerate(repos)]


def _has_overlap(combinations):
    # checks if a combination appears in more than one repository
    for i, repo_combination in enumerate(combinations):
        all_combinations_but_current_repo = set(itertools.chain(*[c for j, c in enumerate(combinations) if i != j]))
        if len(all_combinations_but_current_repo.intersection(repo_combination)) != 0:
            return True
    return False


def _combination_mapping_to_repo_index(combinations):
    # returns a dictionary mapping each (dataset, fold, config) to the repository index
    return {
        (dataset, fold, config): repo_index
        for repo_index, repo_combinations in enumerate(combinations)
        for (dataset, fold, config) in repo_combinations
    }

if __name__ == '__main__':
    repo1 = load_repository("D244_F3_C1530_30", cache=True)
    repo2 = load_repo_artificial()
    combinations = _generate_dataset_fold_config_combinations([repo1, repo2])
    print(combinations)

    mapping = _combination_mapping_to_repo_index(combinations)
    print(mapping)

    print(_has_overlap(combinations))

    combinations = _generate_dataset_fold_config_combinations([repo2, repo2])
    print(_has_overlap(combinations))
```

Reply:
Great suggestion! I adapted this and pushed an update implementing it